src/SAWScript/MGU.hs

```diff
@@ -554,36 +646,64 @@ inferField m (n,e) = do
   (e',t) <- inferE (m,e)
   return ((n,e'),(n,t))
 
-bindSchema :: Located Name -> Schema -> TI a -> TI a
-bindSchema n s m = TI $ local (\ro -> ro { typeEnv = M.insert n s $ typeEnv ro })
-  $ unTI m
+-- wrap m with a type for x
+withVar :: Located Name -> Schema -> TI a -> TI a
+withVar x s m =
+  TI $ local (\ro -> ro { varEnv = M.insert x s $ varEnv ro }) $ unTI m
```

Comment:
It's to make it clearer how they work. Because the environment lives in the "readonly" part of the pass context, you can only change it by wrapping a nested computation with a new readonly value. So basically this one runs the action ```m``` with a binding for the variable.

I at least fairly strongly expect something named ```bind``` to _update_ the environment.

As I noted in a comment further up, I think trying to use Reader for the environments was a mistake and this could all be simplified a lot by changing that, but I didn't want to tackle that in the middle of the other changes.

Reply:
OK. Personally, I am not bothered by using the word "bind" to refer to running a nested read-only computation using `local`. That being said, I don't feel too strongly about this.

Reply:
I think fixing the problems with the interpreter is going to demand we stop using ```Reader``` this way (basically, the interpreter does need to be able to check one statement at a time, and this _does_ need to update the environment, Currently that happens differently outside and inside the typechecker, and I think it's become clear that this shouldn't be allowed.

At that point we can change these all back to ```bind``` :-)